### PR TITLE
Adding Golomb Compressed Sets

### PR DESCRIPTION
**Pull request by Daniel-Liu-c0deb0t**
## Description
Added support for efficient Golomb Compressed Sets (GCS). This has quite a lot of changes, including a more complex protobuf type to handle both Bloom Filters and GCS, new methods for bulk construction and intersection of GCS and Bloom Filters, and updated tests.

The general idea is to use Golomb encoding, which represents a number `n` by dividing a certain power-of-two (which we refer to as `div`). The integer `n / div` is represented in unary, while the remainder `n % div` is represented in binary. After hashing many elements and sorting the hashes, the differences between consecutive elements form a geometric distribution. To get the best `div` value, we can calculate the mean of the differences and approximate the median, which we use as the `div` value. The `count trailing bits` compiler intrinsic is used for efficiency.

**Speed benchmark**: on my machine, GCS is slightly *faster* than Bloom Filters on the current PSI protocol, despite GCS requiring extra sorting operations. This could be due to bad cache locality for Bloom Filters, which require random accesses. For random access of a few elements (not bulk insertions/intersections), Bloom Filters should still be faster.

**Space usage benchmark (length in bytes)**: listed below. GCS leads to a 25-30% space reduction compared to Bloom Filters. Getting significantly more space savings with other methods should be pretty difficult. The lower memory usage should be very good for network transfer in the asymmetric case, as the server set is much larger than the client set. The code is implemented so that the client can decompress the GCS on the fly, to save memory.
```
Max elements: 10000, FPR: 1e-06, div: 19
Naive length (8 bytes per element): 80000
Bloom filter length: 35944
GCS length: 26816
Max elements: 10000, FPR: 1e-07, div: 23
Naive length (8 bytes per element): 80000
Bloom filter length: 41935
GCS length: 30953
Max elements: 10000, FPR: 1e-08, div: 26
Naive length (8 bytes per element): 80000
Bloom filter length: 47926
GCS length: 35058
Max elements: 10000, FPR: 1e-09, div: 29
Naive length (8 bytes per element): 80000
Bloom filter length: 53916
GCS length: 39260
Max elements: 10000, FPR: 1e-10, div: 33
Naive length (8 bytes per element): 80000
Bloom filter length: 59907
GCS length: 43417
Max elements: 10000, FPR: 1e-11, div: 36
Naive length (8 bytes per element): 80000
Bloom filter length: 65898
GCS length: 47516
Max elements: 10000, FPR: 1e-12, div: 39
Naive length (8 bytes per element): 80000
Bloom filter length: 71888
GCS length: 51704
```

## Affected Dependencies
N/A

## How has this been tested?
Extra tests were added.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

**Comment by Daniel-Liu-c0deb0t** [comment]
Note: this is my first time using C++ for a nontrivial program, so please give me feedback!

**Comment by Daniel-Liu-c0deb0t** [comment]
I think I will need some help with the JS bindings. There are a lot of hardcoded test cases that need to be regenerated due to the changes to the ServerSetup protobuf.

**Comment by s0l0ist** [comment]
> I think I will need some help with the JS bindings. There are a lot of hardcoded test cases that need to be regenerated due to the changes to the ServerSetup protobuf.

I can help with that part :)

File: private_set_intersection/cpp/psi_server.h [hunk 1]
```diff
@@ -67,7 +69,8 @@ class PsiServer {
   // Returns INTERNAL if encryption fails.
   StatusOr<psi_proto::ServerSetup> CreateSetupMessage(
       double fpr, int64_t num_client_inputs,
-      absl::Span<const std::string> inputs) const;
+      absl::Span<const std::string> inputs,
+      DataStructure ds = DataStructure::GCS) const;
```
Comment (s0l0ist) [inline private_set_intersection/cpp/psi_server.h:73]: Should we make the default `BloomFilter` instead?
```suggestion
      DataStructure ds = DataStructure::BloomFilter) const;
```

**Pull request merged**